hw/xbox/xid.c

```diff
@@ -327,28 +336,35 @@ static void usb_xid_handle_control(USBDevice *dev, USBPacket *p,
     /* HID requests */
     case ClassInterfaceRequest | HID_GET_REPORT:
         DPRINTF("xid GET_REPORT 0x%x\n", value);
-        if (value == 0x100) { /* input */
-            assert(s->in_state.bLength <= length);
-            // s->in_state.bReportId++; /* FIXME: I'm not sure if bReportId is just a counter */
-            memcpy(data, &s->in_state, s->in_state.bLength);
-            p->actual_length = s->in_state.bLength;
+        if (value == 0x0100) { /* input */
+            if (length <= s->in_state.bLength) {
+                memcpy(data, &s->in_state, s->in_state.bLength);
+                p->actual_length = length;
+            } else {
+                p->status = USB_RET_STALL;
+            }
         } else {
+            p->status = USB_RET_STALL;
             assert(false);
         }
         break;
     case ClassInterfaceOutRequest | HID_SET_REPORT:
         DPRINTF("xid SET_REPORT 0x%x\n", value);
-        if (value == 0x200) { /* output */
+        if (value == 0x0200) { /* output */
             /* Read length, then the entire packet */
-            memcpy(&s->out_state, data, sizeof(s->out_state));
-            assert(s->out_state.length == sizeof(s->out_state));
-            assert(s->out_state.length <= length);
-            //FIXME: Check actuator endianess
-            DPRINTF("Set rumble power to 0x%x, 0x%x\n",
-                    s->out_state.left_actuator_strength,
-                    s->out_state.right_actuator_strength);
-            p->actual_length = s->out_state.length;
+            if (length == s->out_state.length) {
+                memcpy(&s->out_state, data, sizeof(s->out_state));
+
+                /* FIXME: This should also be a STALL */
+                assert(s->out_state.length == sizeof(s->out_state));
+
+                p->actual_length = length;
+            } else {
+                p->status = USB_RET_STALL;
+            }
+            update_force_feedback(s);
```

Comment:
Should this be moved inside the true branch of above condition?

Reply:
I also thought about it but figured it doesn't matter.
This is also more extensible for handling different packet length (which might require different code paths) or if we find out that we have to stop rumble on STALL.